File: src/scheme/evaluator/define.ts
import { type Thunk, type SchemeData, SchemeCont, SchemeSym, SchemeList, SchemeProc } from '../parser/data'
import type { Env } from '../env'
import type { IEvaluator, Evaluator } from './index'
import { assert } from '../utils'

/**
 * 语法：
 * 1.定义变量或函数：
 * (define vhello "Hello world")
 * 2.短形式定义函数：
 * (define (hello name) (string-append "Hello " name "!"))
 */

export default class DefineEvaluator implements IEvaluator {
  constructor(private evaluator: Evaluator) {}

  public matches(value: string): boolean {
    return value === 'define'
  }

  public evaluate(node: SchemeList, env: Env, cont: SchemeCont): Thunk {
    const varNode = node.cadr()
    const bodyNode = node.cddr()

    // 定义变量或者函数
    if (SchemeSym.matches(varNode)) {
      return this.evaluator.evaluate(bodyNode, env, new SchemeCont((data: SchemeData) => () => {
        env.define(varNode.value, data)
        return cont.call(data)
      }))
    }

    // 定义函数
    if (SchemeList.matches(varNode)) {
      const name = SchemeSym.cast(varNode.car()).value
      const proc = new SchemeProc(name, varNode.cdr(), bodyNode, env)
      env.define(name, proc)
      return cont.call(proc)
    }

    assert(false, 'Error: define clause evaluate error!')
  }
}

File: packages/interpreter-rust/src/evaluator/define.rs
use std::cell::RefCell;
use std::rc::Rc;

use crate::env::Env;
use crate::evaluator::{Evaluator, IEvaluator};
use crate::parser::{SchemeCont, SchemeData, SchemeExp};

pub struct DefineEvaluator;

impl IEvaluator for DefineEvaluator {
  fn can_match(&self, data: &SchemeExp) -> bool {
    true
  }
  fn evaluate(
    &self,
    data: &SchemeExp,
    env: &Rc<RefCell<Env>>,
    cont: &SchemeCont,
    base_evaluator: &Evaluator,
  ) -> SchemeData {
    SchemeData::Nil
  }
}
